source/common/upstream/conn_pool_map_impl.h

```diff
@@ -63,5 +75,38 @@ void ConnPoolMap<KEY_TYPE, POOL_TYPE>::drainConnections() {
     pool_pair.second->drainConnections();
   }
 }
+
+template <typename KEY_TYPE, typename POOL_TYPE>
+void ConnPoolMap<KEY_TYPE, POOL_TYPE>::poolDrained(const KEY_TYPE& key) {
+  if (adding_pool_) {
+    drained_pools_.push_back(key);
+  } else {
+    active_pools_.erase(key);
+  }
+}
+
+template <typename KEY_TYPE, typename POOL_TYPE>
+bool ConnPoolMap<KEY_TYPE, POOL_TYPE>::cleanDrainedPools() {
+  if (drained_pools_.empty()) {
+    return false;
+  }
+
+  for (auto& key : drained_pools_) {
+    active_pools_.erase(key);
+  }
+
+  drained_pools_.clear();
+  return true;
+}
+
+template <typename KEY_TYPE, typename POOL_TYPE>
+void ConnPoolMap<KEY_TYPE, POOL_TYPE>::registerFullCallbacks() {
```

Comment:
This is only called from one place, right? Can you inline it there and add more comments on what the logic is doing? It's not immediately obvious from a quick read.

Reply:
I typically do this (take loops, conditions, etc) out into their own functions as a method of reducing cyclomatic complexity and depth-of-nesting -- I find it leads to more maintainable and readable code. If you'd prefer it inlined I can certainly do that. In fact, it was inlined originally. I only moved it out on a second pass. :)

Either way, it can definitely use more comments.

Reply:
Either way is fine, just please add some more comments and I can take another pass.